src/textinput/TextInput.ts

```diff
@@ -0,0 +1,37 @@
+import { WidgetBase } from '@dojo/widget-core/WidgetBase';
+import { ThemeableMixin, ThemeableProperties, theme } from '@dojo/widget-core/mixins/Themeable';
+import { FormLabelMixin, FormLabelMixinProperties } from '@dojo/widget-core/mixins/FormLabel';
+import { v } from '@dojo/widget-core/d';
+import * as css from './styles/textinput.css';
+
+/**
+ * @type TextInputProperties
+ *
+ * Properties that can be set on a TextInput component
+ *
+ * @property type			Input type, e.g. text, email, tel, etc.
+ * @property onInput	Called when the node's 'input' event is fired
+ */
+export interface TextInputProperties extends ThemeableProperties, FormLabelMixinProperties {
+	type?: string;
+	onInput?(event: Event): void;
+}
+
+@theme(css)
+export default class TextInput extends ThemeableMixin(FormLabelMixin(WidgetBase))<TextInputProperties> {
+	onInput(event: Event) {
+		this.properties.onInput && this.properties.onInput(event);
+	}
+
+	render() {
```

Comment:
@bitpshr the problem with the approach you are suggesting is that it is assuming that every widget wants to update it's value `onInput`. It feels like you're trying to make the data flow bi-directional such that the parent knows the value of the input at all times which goes against out uni-directional reactive design principles. Even without triggering a render, there will be an overhead involved with firing an `onInput` function on every key stroke. The `value` could be accessed in a number of ways depending on the use case; `onFocus`, `onBlur`, `onSubmit` to name a few.  

Also worth considering that external state can be `async`  which would introduce even more problems as you cannot guarantee the order of the callbacks thus the value could easily be corrupted. Using the `initialValue` approach keeps the concerns separate and leaves widget authors / app creators able to get the `value` via whichever callback / event they desire. The uncontrolled approach keeps the single source of truth in the dom.